GetFullName in TypeExtensions.cs not handling arrays correctly

### What You Are Seeing?
When creating a CakeMethodAlias method with a `params System.Linq.Expressions.Expression<Func<string,string>> expressions` parameter, the resulting Roslyn compilation fails.

```
Error: Roslyn.Compilers.CompilationErrorException: (1918,131): error CS0231: A params parameter must be the last parameter in a formal parameter list
   at Roslyn.Scripting.CommonScriptEngine.CompilationError(DiagnosticBag localDiagnostics, DiagnosticBag diagnostics)
   at Roslyn.Scripting.CommonScriptEngine.Compile(String code, String path, DiagnosticBag diagnostics, Session session, Type delegateType, Type returnType, CancellationToken cancellationToken, Boolean
 isInteractive, Boolean isExecute, CommonCompilation& compilation, Delegate& factory)
   at Roslyn.Scripting.CommonScriptEngine.Execute[T](String code, String path, DiagnosticBag diagnostics, Session session, Boolean isInteractive)
   at Roslyn.Scripting.Session.Execute(String code)
   at Cake.Scripting.Roslyn.Stable.DefaultRoslynScriptSession.Execute(Script script) in C:\git\cake\src\Cake\Scripting\Roslyn\Stable\DefaultRoslynScriptSession.cs:line 31
   at Cake.Core.Scripting.ScriptRunner.Run(IScriptHost host, FilePath scriptPath, IDictionary`2 arguments) in C:\git\cake\src\Cake.Core\Scripting\ScriptRunner.cs:line 189
   at Cake.Commands.BuildCommand.Execute(CakeOptions options) in C:\git\cake\src\Cake\Commands\BuildCommand.cs:line 33
   at Cake.CakeApplication.Run(CakeOptions options) in C:\git\cake\src\Cake\CakeApplication.cs:line 46
   at Cake.Program.Main() in C:\git\cake\src\Cake\Program.cs:line 69
```
### What is Expected?

I would expect the compilation to work as expected.

### What version of Cake are you using?

0.14

### Are you running on a 32 or 64 bit system?
x64

### What environment are you running on?  Windows? Linux? Mac?
Windows

### Are you running on a CI Server?  If so, which one?
No

### How Did You Get This To Happen? (Steps to Reproduce)

Author a Cake add-in with the following method signature - actual implementation is obfuscated, it's the method signature that's causing the issue.
```cs
[CakeMethodAlias]
public static object DoSomething(this ICakeContext context, FilePath path, params Expression<Func<string, string>>[] properties)
{
    if (context == null)
    {
        throw new ArgumentNullException(nameof(context));
    }

    // implementation ...

    throw new NotImplementedException();
}
```
### Output Log
<!--
When including the log information, please ensure you have run the command with --verbosity=diagnostic. It provides important information for determining an issue.

If running Cake.exe directly, the parameter is passed in as --verbosity=diagnostic
If running Cake via the bootstrapper, the parameter is -Verbosity Diagnostic

- Make sure there is no sensitive data shared.
- We need ALL output, not just what you may believe is relevant.

In order to rule out potential issues with the compilation of your Cake script, it would be very helpful if you could try running the script with some additional parameters.

The first, when compiling on Windows, would be to use the -Mono flag.  This can be passed into both Cake.exe and the Bootstrapper.  This will allow verification that it isn't a problem specific to one compiler or the other.

The second, would be to use the latest version of Roslyn.  This can be done using the -Experimental flag, which again can be passed to both Cake.exe and the bootstrapper.

If you can provide the output from the above when submitting an issue, this would be tremendously useful!
-->

https://gist.github.com/kcamp/b08e1b5527baf2216a8005628652dbef

Further investigation shows that the underlying code being serialized for Roslyn in RoslynCodeGenerator.cs is malformed.

~~~cs
public object DoSomething(Cake.Core.IO.FilePath path, params System.Linq.Expressions.Expression`1[[System.Func`2[[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]][] properties)
{
    return My.NameSpace.ProjectAliases.DoSomething(Context, path, properties);
}
~~~
<!--
TEMPLATE FOR FEATURE REQUESTS:

It's a blank slate, have fun!
-->

--- a/src/Cake.Core.Tests/Cake.Core.Tests.csproj
+++ b/src/Cake.Core.Tests/Cake.Core.Tests.csproj
@@ -163,6 +163,10 @@
     <EmbeddedResource Include="Unit\Scripting\CodeGen\Expected\Properties\Cached_Obsolete_ImplicitWarning_NoMessage" />
     <EmbeddedResource Include="Unit\Scripting\CodeGen\Expected\Properties\Cached_Obsolete_ExplicitWarning_WithMessage" />
     <EmbeddedResource Include="Unit\Scripting\CodeGen\Expected\Properties\Cached_Obsolete_ExplicitError_WithMessage" />
+    <EmbeddedResource Include="Unit\Scripting\CodeGen\Expected\Methods\NonGeneric_ExtensionMethodWithGenericExpressionParameter" />
+    <EmbeddedResource Include="Unit\Scripting\CodeGen\Expected\Methods\NonGeneric_ExtensionMethodWithGenericExpressionArrayParameter" />
+    <EmbeddedResource Include="Unit\Scripting\CodeGen\Expected\Methods\NonGeneric_ExtensionMethodWithGenericExpressionParamsArrayParameter" />
+    <EmbeddedResource Include="Unit\Scripting\CodeGen\Expected\Methods\NonGeneric_ExtensionMethodWithArrayParameter" />
   </ItemGroup>
   <ItemGroup>
     <ProjectReference Include="..\Cake.Core\Cake.Core.csproj">

--- a/src/Cake.Core.Tests/Data/MethodAliasGeneratorData.cs
+++ b/src/Cake.Core.Tests/Data/MethodAliasGeneratorData.cs
@@ -3,6 +3,7 @@
 // See the LICENSE file in the project root for more information.
 using System;
 using System.Diagnostics;
+using System.Linq.Expressions;
 using Cake.Core.Annotations;
 
 namespace Cake.Core.Tests.Data
@@ -39,6 +40,30 @@ namespace Cake.Core.Tests.Data
 
         [CakeMethodAlias]
         public static void NonGeneric_ExtensionMethodWithGenericParameter(this ICakeContext context, Action<int> value)
+        {
+            throw new NotImplementedException();
+        }
+
+        [CakeMethodAlias]
+        public static void NonGeneric_ExtensionMethodWithGenericExpressionParameter(this ICakeContext context, Expression<Func<string, string>> expression)
+        {
+            throw new NotImplementedException();
+        }
+
+        [CakeMethodAlias]
+        public static void NonGeneric_ExtensionMethodWithGenericExpressionArrayParameter(this ICakeContext context, Expression<Func<string, string>>[] expression)
+        {
+            throw new NotImplementedException();
+        }
+
+        [CakeMethodAlias]
+        public static void NonGeneric_ExtensionMethodWithGenericExpressionParamsArrayParameter(this ICakeContext context, params Expression<Func<string, string>>[] expression)
+        {
+            throw new NotImplementedException();
+        }
+
+        [CakeMethodAlias]
+        public static void NonGeneric_ExtensionMethodWithArrayParameter(this ICakeContext context, string[] values)
         {
             throw new NotImplementedException();
         }

--- a/src/Cake.Core.Tests/Unit/Scripting/CodeGen/Expected/Methods/NonGeneric_ExtensionMethodWithArrayParameter
+++ b/src/Cake.Core.Tests/Unit/Scripting/CodeGen/Expected/Methods/NonGeneric_ExtensionMethodWithArrayParameter
@@ -1,0 +1,4 @@
+﻿public void NonGeneric_ExtensionMethodWithArrayParameter(System.String[] values)
+{
+    Cake.Core.Tests.Data.MethodAliasGeneratorData.NonGeneric_ExtensionMethodWithArrayParameter(Context, values);
+}

--- a/src/Cake.Core.Tests/Unit/Scripting/CodeGen/Expected/Methods/NonGeneric_ExtensionMethodWithGenericExpressionArrayParameter
+++ b/src/Cake.Core.Tests/Unit/Scripting/CodeGen/Expected/Methods/NonGeneric_ExtensionMethodWithGenericExpressionArrayParameter
@@ -1,0 +1,4 @@
+﻿public void NonGeneric_ExtensionMethodWithGenericExpressionArrayParameter(System.Linq.Expressions.Expression<System.Func<System.String, System.String>>[] expression)
+{
+    Cake.Core.Tests.Data.MethodAliasGeneratorData.NonGeneric_ExtensionMethodWithGenericExpressionArrayParameter(Context, expression);
+}

--- a/src/Cake.Core.Tests/Unit/Scripting/CodeGen/Expected/Methods/NonGeneric_ExtensionMethodWithGenericExpressionParameter
+++ b/src/Cake.Core.Tests/Unit/Scripting/CodeGen/Expected/Methods/NonGeneric_ExtensionMethodWithGenericExpressionParameter
@@ -1,0 +1,4 @@
+﻿public void NonGeneric_ExtensionMethodWithGenericExpressionParameter(System.Linq.Expressions.Expression<System.Func<System.String, System.String>> expression)
+{
+    Cake.Core.Tests.Data.MethodAliasGeneratorData.NonGeneric_ExtensionMethodWithGenericExpressionParameter(Context, expression);
+}

--- a/src/Cake.Core.Tests/Unit/Scripting/CodeGen/Expected/Methods/NonGeneric_ExtensionMethodWithGenericExpressionParamsArrayParameter
+++ b/src/Cake.Core.Tests/Unit/Scripting/CodeGen/Expected/Methods/NonGeneric_ExtensionMethodWithGenericExpressionParamsArrayParameter
@@ -1,0 +1,4 @@
+﻿public void NonGeneric_ExtensionMethodWithGenericExpressionParamsArrayParameter(params System.Linq.Expressions.Expression<System.Func<System.String, System.String>>[] expression)
+{
+    Cake.Core.Tests.Data.MethodAliasGeneratorData.NonGeneric_ExtensionMethodWithGenericExpressionParamsArrayParameter(Context, expression);
+}

--- a/src/Cake.Core.Tests/Unit/Scripting/CodeGen/MethodAliasGeneratorTests.cs
+++ b/src/Cake.Core.Tests/Unit/Scripting/CodeGen/MethodAliasGeneratorTests.cs
@@ -32,6 +32,9 @@ namespace Cake.Core.Tests.Unit.Scripting.CodeGen
             [InlineData("NonGeneric_ExtensionMethodWithNoParameters")]
             [InlineData("NonGeneric_ExtensionMethodWithParameter")]
             [InlineData("NonGeneric_ExtensionMethodWithGenericParameter")]
+            [InlineData("NonGeneric_ExtensionMethodWithGenericExpressionParameter")]
+            [InlineData("NonGeneric_ExtensionMethodWithGenericExpressionArrayParameter")]
+            [InlineData("NonGeneric_ExtensionMethodWithGenericExpressionParamsArrayParameter")]
             [InlineData("NonGeneric_ExtensionMethodWithReturnValue")]
             [InlineData("NonGeneric_ExtensionMethodWithParameterArray")]
             public void Should_Return_Correct_Generated_Code_For_Non_Generic_Methods(string name)

--- a/src/Cake.Core/Extensions/TypeExtensions.cs
+++ b/src/Cake.Core/Extensions/TypeExtensions.cs
@@ -47,6 +47,12 @@ namespace Cake.Core
             {
                 return type.Name;
             }
+
+            if (type.IsArray)
+            {
+                return type.GetElementType().GetFullName(includeNamespace) + "[]";
+            }
+
             Type genericType;
             return type.IsGenericType(out genericType)
                 ? GetGenericTypeName(genericType, includeNamespace)

